pkg/helper/webconsole.go

```diff
@@ -0,0 +1,144 @@
+
+// ConsoleLinkNamePrefix is the prefix applied to Console link for system master
+const ConsoleLinkNamePrefix = "system-master-link"
+
+// ConsoleLinkSupported checks if a ConsoleLink CRD exists
+func ConsoleLinkSupported() error {
+	gvk := schema.GroupVersionKind{Group: "console.openshift.io", Version: "v1", Kind: "ConsoleLink"}
+	return kubernetes.CustomResourceDefinitionExists(gvk)
+}
+
+// ReconcileConsoleLink creates/deletes/updates a ConsoleLink based on the Route
+func ReconcileConsoleLink(ctx context.Context, c client.Client, req *reconcile.Request) error {
+	route := getRoute(ctx, c, req)
+	if route == nil {
+		if err := removeConsoleLink(ctx, c, req); err != nil {
+			return err
+		}
+	} else {
+		if err := reconcileConsoleLink(ctx, c, req, route); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getRoute retrieves all the routes in the current namespace and return the desired one.
+func getRoute(ctx context.Context, c client.Client, req *reconcile.Request) *routev1.Route {
+	reader := read.New(c).WithNamespace(req.Namespace)
+	deployedRoutes, err := reader.List(&routev1.RouteList{})
+	if err == nil {
+		for _, route := range deployedRoutes { //capture targeted route object
+			realr := route.(*routev1.Route)
+			if strings.Compare(realr.Spec.To.Name, "system-master") == 0 {
+				return realr
+			}
+		}
+	}
+	return nil
+}
+
+// reconcileConsoleLink create a ConsoleLink for the Route, or update the existing ConsoleLink if it exists.
+// return error if update or creation could not be done
+func reconcileConsoleLink(ctx context.Context, c client.Client, req *reconcile.Request, route *routev1.Route) error {
+	consoleLinkName := fmt.Sprintf("%s-%s", ConsoleLinkNamePrefix, req.Namespace)
+	consoleLink := &consolev1.ConsoleLink{}
+	err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+	if err != nil && apierrors.IsNotFound(err) {
+		consoleLink = getConsoleLink(consoleLinkName, route, req)
+		if err := c.Create(ctx, consoleLink); err != nil {
+			err = fmt.Errorf("failed to create consolelink %s %w", consoleLinkName, err)
+			logger.Error(err)
+			return err
+		}
+		logger.Info("consolelink has been created:", consoleLinkName)
+	} else if err == nil && consoleLink != nil {
+		if err = updateConsoleLink(ctx, route, consoleLink, c); err != nil {
+			return err
+		}
+	} else {
+		err = fmt.Errorf("failed to retrieve consolelink %s %w", consoleLinkName, err)
+		logger.Error(err)
+		return err
+	}
+	return nil
+}
+
+// removeConsoleLink removes a ConsoleLink object if it exists
+// error is returned if the deletion is failed.
+func removeConsoleLink(ctx context.Context, c client.Client, req *reconcile.Request) error {
+	consoleLink := &consolev1.ConsoleLink{}
+	consoleLinkName := fmt.Sprintf("%s-%s", ConsoleLinkNamePrefix, req.Namespace)
+	err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+	if err == nil && consoleLink != nil {
```

Comment:
This checking seems odd.

If err is nil ConsoleLink shouldn't be nil by definition right? :thinking:

Reply:
Additionally, if the error is not nil and it is not the isnotfound error then the error should be returned so it can be retried